speechbrain/utils/metric_stats.py

```diff
@@ -358,10 +374,235 @@ def write_stats(self, filestream):
         print_alignments(self.scores, filestream)
 
 
-class BinaryMetricStats(MetricStats):
-    """Tracks binary metrics, such as precision, recall, F1, EER, etc.
+class WeightedErrorRateStats(MetricStats):
+    """Metric that reweighs the WER from :class:`~ErrorRateStats` with any
+    chosen method. This does not edit the sequence of found edits
+    (insertion/deletion/substitution) but multiplies their impact on the metric
+    by a value between 0 and 1 as returned by the cost function.
+
+    Arguments
+    ---------
+    base_stats : ErrorRateStats
+        The base WER calculator to use.
+    cost_function : Callable[[str, Optional[str], Optional[str]], float]
+        Cost function of signature `fn(edit_symbol, a, b) -> float`, where the
+        returned value, between 0 and 1, is the weight that should be assigned
+        to a particular edit in the weighted WER calculation.
+        In the case of insertions and deletions, either of `a` or `b` may be
+        `None`. In the case of substitutions, `a` and `b` will never be `None`.
+    weight_name : str
+        Prefix to be prepended to each metric name (e.g. `xxx_wer`)
     """
 
+    def __init__(
+        self,
+        base_stats: ErrorRateStats,
+        cost_function: Callable[[str, Optional[str], Optional[str]], float],
+        weight_name: str = "weighted",
+    ):
+        self.clear()
+        self.base_stats = base_stats
+        self.cost_function = cost_function
+        self.weight_name = weight_name
+
+    def append(self, *args, **kwargs):
+        """Append function, which should **NOT** be used for the weighted error
+        rate stats. Please append to the specified `base_stats` instead.
+
+        Arguments
+        ---------
+        *args
+            Ignored.
+        **kwargs
+            Ignored.
+        """
+
+        raise ValueError(
+            "Cannot append to a WeightedErrorRateStats. "
+            "You should only append to the base ErrorRateStats."
+        )
+
+    def summarize(self, field=None):
+        """Returns a dict containing some detailed WER statistics after weighting
+        every edit with a weight determined by `cost_function` (returning `0.0`
+        for no error, `1.0` for the default error behavior, and anything in
+        between).
+
+        See :meth:`~ErrorRateStats.summarize`."""
```

Comment:
I would say we need to imrpove the docstring to better reflect input / output variables